packages/puppeteer-core/src/common/Input.ts

```diff
@@ -438,90 +513,178 @@ export class Mouse {
-    if (delay !== null) {
-      await new Promise(f => {
-        return setTimeout(f, delay);
+    const {steps = 1} = options;
+    const from = this.#state.position;
+    const to = {x, y};
+    for (let i = 1; i <= steps; i++) {
+      await this.#withTransaction(updateState => {
+        updateState({
+          position: {
+            x: from.x + (to.x - from.x) * (i / steps),
+            y: from.y + (to.y - from.y) * (i / steps),
+          },
+        });
+        const {buttons, position} = this.#state;
+        return this.#client.send('Input.dispatchMouseEvent', {
+          type: 'mouseMoved',
+          modifiers: this.#keyboard._modifiers,
+          buttons,
+          // This should always be 0 (i.e. 'left'). See
+          // https://w3c.github.io/uievents/#event-type-mousemove
+          button: MouseButton.Left,
+          ...position,
+        });
       });
     }
-    await this.up(options);
   }
 
   /**
-   * Dispatches a `mousedown` event.
-   * @param options - Optional `MouseOptions`.
+   * Presses the mouse.
+   *
+   * @param options - Options to configure behavior.
    */
   async down(options: MouseOptions = {}): Promise<void> {
-    const {button = 'left', clickCount = 1} = options;
-    this.#button = button;
-    await this.#client.send('Input.dispatchMouseEvent', {
-      type: 'mousePressed',
-      button,
-      x: this.#x,
-      y: this.#y,
-      modifiers: this.#keyboard._modifiers,
-      clickCount,
+    const {button = MouseButton.Left, clickCount = 1} = options;
+    const flag = getFlag(button);
+    if (!flag) {
+      throw new Error(`Unsupported mouse button: ${button}`);
+    }
+    if (this.#state.buttons & flag) {
+      throw new Error(`'${button}' is already pressed.`);
+    }
+    await this.#withTransaction(updateState => {
+      updateState({buttons: this.#state.buttons | flag});
+      const {buttons, position} = this.#state;
+      return this.#client.send('Input.dispatchMouseEvent', {
+        type: 'mousePressed',
+        modifiers: this.#keyboard._modifiers,
+        clickCount,
+        buttons,
+        button,
+        ...position,
+      });
     });
   }
 
   /**
-   * Dispatches a `mouseup` event.
-   * @param options - Optional `MouseOptions`.
+   * Releases the mouse.
+   *
+   * @param options - Options to configure behavior.
    */
   async up(options: MouseOptions = {}): Promise<void> {
-    const {button = 'left', clickCount = 1} = options;
-    this.#button = 'none';
-    await this.#client.send('Input.dispatchMouseEvent', {
-      type: 'mouseReleased',
-      button,
-      x: this.#x,
-      y: this.#y,
-      modifiers: this.#keyboard._modifiers,
-      clickCount,
+    const {button = MouseButton.Left, clickCount = 1} = options;
+    const flag = getFlag(button);
+    if (!flag) {
+      throw new Error(`Unsupported mouse button: ${button}`);
+    }
+    if (!(this.#state.buttons & flag)) {
+      throw new Error(`'${button}' is not pressed.`);
```

Comment:
Previously, if you called `up` and the button wasn't pressed, this function would not throw an error.

Should this have been a breaking change?

For example, the testing library I use has a `resetMouse` function which calls `up` on all buttons to make sure the mouse is in it's initial state and after this change using that function throws an error.

Reply:
@jrandolf

Reply:
So are you requesting the `up` button reset to a no-op?

Reply:
My main goal was just to clarify the breaking behavior of the change since I'm not an expert in this code base. 2 weeks later there might not be much to do except maybe updating the release notes so that it's easier for users to find the breaking change.

Perhaps it would be useful if the documentation should be updated to clarify the constraints on calling the methods.

As to the actual design decision, I think it comes down to user expectations. If we think most users calling `up` without calling `down` are doing it by accident then throwing an error is probably better. If not, then maybe a warning would be better (or nothing).

Since users who want to unconditionally reset mouse can use `try...catch` to avoid major problems with the error, it's fine to inconvenience them if they're in the minority imo.

Reply:
@jrandolf let's improve documentation around this (by calling out the new error more explicitly in the PR description and updating the method docs)?

Reply:
This indeed seems to have broken a package of ours: https://github.com/modernweb-dev/web/issues/2262

Apparently now: "it's an error to call up if the mouse button is not actually pressed"

The code in our package looks as follows: https://github.com/modernweb-dev/web/blob/e0d362fcdbedf6a72a6c78ebc2e3a3cffa922dbc/packages/test-runner-commands/src/sendMousePlugin.ts#L170

Could a maintainer of puppeteer provide some guidance on what would be the fix for this? Apparently we just need to "actually press" the mouse button? What would that look like?

Reply:
```
await page.mouse.down({ button: 'left' });
```

should be called before up:

```
await page.mouse.up({ button: 'left' });
```

it looks like you only send up calls without the down calls. Alternatively, it seems you can use `click` to have down/up paired up correctly.

Reply:
Their `up` calls are being used to just reset the mouse state. WDYT about having a `reset` function on the mouse? This way, the user doesn't need to keep track of the mouse state to reset it.

Reply:
@jrandolf should we have some sort of reset mechanism for the mouse/keyboard? I can see the use case for tests where it is difficult to know if the tests left the keyboard/mouse in the right state.

Reply:
Could you ping me when this is implemented and released? Seems like a `reset` function would be great here indeed :)

Reply:
See https://github.com/puppeteer/puppeteer/pull/10340